File: python/invaders_game/game.py
"""ゲーム全体の状態管理・衝突判定・メインループを担当するモジュール。"""

from typing import List

import pygame

from barrier import Barrier
from bullet import Bullet
from enemy import EnemySwarm
from player import Player


class Game:
    """ウィンドウ・入力・更新・描画を束ねるゲーム本体。"""

    WIDTH = 800
    HEIGHT = 600
    FPS = 60
    # レベル0〜3の4段階（最後のステージをクリアするとゲームクリア）
    MAX_LEVEL = 3
    INITIAL_LIVES = 3
    PLAYER_SHOOT_COOLDOWN_MS = 300
    ENEMY_SHOOT_INTERVAL_BASE_MS = 950

    def __init__(self) -> None:
        pygame.init()
        pygame.display.set_caption("スペースインベーダー風")
        self.screen = pygame.display.set_mode((self.WIDTH, self.HEIGHT))
        self.clock = pygame.time.Clock()
        self.font = pygame.font.Font(None, 28)
        self.font_large = pygame.font.Font(None, 56)
        self.font_hint = pygame.font.Font(None, 22)

        self.ground_y = self.HEIGHT - 48
        # 敵の下端がこのY以上なら自機ライン到達とみなす
        self.invasion_y = self.ground_y - Player.HEIGHT - 8

        self.score = 0
        self.level = 0
        self.lives = self.INITIAL_LIVES
        # playing / game_over / game_clear
        self.state = "playing"

        self.player: Player
        self.swarm: EnemySwarm
        self.barriers: List[Barrier]
        self.player_bullets: List[Bullet] = []
        self.enemy_bullets: List[Bullet] = []

        self._shoot_cooldown_ms = 0
        self._enemy_shoot_accum_ms = 0

        self._spawn_stage()

    def _spawn_stage(self) -> None:
        """現在のレベルに応じてステージ（敵・バリア・自機位置）を生成する。"""
        self.player = Player(self.WIDTH, self.ground_y)
        self.swarm = EnemySwarm(self.WIDTH, self.level)
        barrier_top = self.ground_y - 110
        # 画面幅に沿って4基を等間隔配置
        xs = [int(self.WIDTH * (2 * i + 1) / 10) for i in range(4)]
        self.barriers = [Barrier(cx, barrier_top) for cx in xs]
        self.player_bullets.clear()
        self.enemy_bullets.clear()
        self._shoot_cooldown_ms = 0
        self._enemy_shoot_accum_ms = 0

    def _restart_from_beginning(self) -> None:
        """スコア・レベル・残機を初期化して最初からやり直す。"""
        self.score = 0
        self.level = 0
        self.lives = self.INITIAL_LIVES
        self.state = "playing"
        self._spawn_stage()

    def _enemy_shoot_interval_ms(self) -> int:
        """レベルが上がるほど敵の射撃間隔が短くなる。"""
        return max(380, self.ENEMY_SHOOT_INTERVAL_BASE_MS - self.level * 140)

    def _try_player_shoot(self, keys: pygame.key.ScancodeWrapper, dt_ms: int) -> None:
        """スペースキーでクールダウン付き射撃。"""
        self._shoot_cooldown_ms = max(0, self._shoot_cooldown_ms - dt_ms)
        if not keys[pygame.K_SPACE] or self._shoot_cooldown_ms > 0:
            return
        speed = -11
        bullet = Bullet(
            float(self.player.rect.centerx),
            float(self.player.rect.top),
            speed,
            from_player=True,
        )
        self.player_bullets.append(bullet)
        self._shoot_cooldown_ms = self.PLAYER_SHOOT_COOLDOWN_MS

    def _enemy_try_shoot(self, dt_ms: int) -> None:
        """ランダムな最前列の敵から下向き弾を発射。"""
        self._enemy_shoot_accum_ms += dt_ms
        if self._enemy_shoot_accum_ms < self._enemy_shoot_interval_ms():
            return
        self._enemy_shoot_accum_ms = 0
        shooter = self.swarm.pick_random_shooter()
        if shooter is None:
            return
        vy = 4.0 + self.level * 1.2
        b = Bullet(
            float(shooter.rect.centerx),
            float(shooter.rect.bottom),
            vy,
            from_player=False,
        )
        self.enemy_bullets.append(b)

    def _update_bullets(self) -> None:
        """弾の移動と画面外削除。"""
        for b in self.player_bullets:
            b.update()
            b.cull_if_off_screen(self.HEIGHT)
        for b in self.enemy_bullets:
            b.update()
            b.cull_if_off_screen(self.HEIGHT)
        self.player_bullets = [b for b in self.player_bullets if b.alive]
        self.enemy_bullets = [b for b in self.enemy_bullets if b.alive]

    def _collide_player_bullets(self) -> None:
        """自機弾とバリア・敵の衝突処理。"""
        for bullet in self.player_bullets:
            if not bullet.alive:
                continue
            for barrier in self.barriers:
                if barrier.collide_bullet(bullet.rect):
                    bullet.alive = False
                    break
            if not bullet.alive:
                continue
            for enemy in self.swarm.enemies:
                if enemy.alive and enemy.rect.colliderect(bullet.rect):
                    enemy.alive = False
                    bullet.alive = False
                    self.score += enemy.score_value
                    break

    def _collide_enemy_bullets(self) -> None:
        """敵弾とバリア・自機の衝突処理。"""
        for bullet in self.enemy_bullets:
            if not bullet.alive:
                continue
            for barrier in self.barriers:
                if barrier.collide_bullet(bullet.rect):
                    bullet.alive = False
                    break
            if not bullet.alive:
                continue
            if bullet.rect.colliderect(self.player.rect):
                bullet.alive = False
                self._on_player_hit()

    def _on_player_hit(self) -> None:
        """自機被弾：残機を減らし、弾をクリア。0ならゲームオーバー。"""
        self.lives -= 1
        self.player.reset_position()
        self.enemy_bullets.clear()
        if self.lives <= 0:
            self.state = "game_over"

    def _check_stage_progress(self) -> None:
        """全滅時に次レベルへ進むか、最終クリアかを判定。"""
        if self.swarm.alive_count() > 0:
            return
        if self.level >= self.MAX_LEVEL:
            self.state = "game_clear"
            return
        self.level += 1
        self._spawn_stage()

    def _check_invasion(self) -> None:
        """敵が自機ラインに達したらゲームオーバー。"""
        if self.swarm.invasion_line_reached(self.invasion_y):
            self.state = "game_over"

    def _update_playing(self, keys: pygame.key.ScancodeWrapper, dt_ms: int) -> None:
        """1フレーム分のプレイ中ロジック。"""
        self.player.update(keys)
        self._try_player_shoot(keys, dt_ms)
        self.swarm.update()
        self._enemy_try_shoot(dt_ms)
        self._update_bullets()
        self._collide_player_bullets()
        self._collide_enemy_bullets()
        # 被弾でゲームオーバーになった場合は侵攻判定・ステージ進行を行わない
        if self.state != "playing":
            return
        self._check_invasion()
        if self.state != "playing":
            return
        self._check_stage_progress()

    def _draw_hud(self) -> None:
        """スコア・残機・レベルを画面上部に表示する。"""
        margin = 8
        score_surf = self.font.render(f"SCORE {self.score}", True, (230, 230, 230))
        lives_surf = self.font.render(f"LIVES {self.lives}", True, (230, 230, 230))
        level_surf = self.font.render(f"LEVEL {self.level}", True, (230, 230, 230))

        self.screen.blit(score_surf, (margin, margin))
        self.screen.blit(lives_surf, (margin + 220, margin))
        self.screen.blit(level_surf, (margin + 400, margin))

    def _draw_overlay(self, title: str, subtitle: str) -> None:
        """ゲームオーバー／クリア時の半透明オーバーレイ。"""
        veil = pygame.Surface((self.WIDTH, self.HEIGHT), pygame.SRCALPHA)
        veil.fill((0, 0, 0, 170))
        self.screen.blit(veil, (0, 0))

        title_surf = self.font_large.render(title, True, (255, 255, 200))
        sub_surf = self.font_hint.render(subtitle, True, (220, 220, 220))

        tr = title_surf.get_rect(center=(self.WIDTH // 2, self.HEIGHT // 2 - 30))
        sr = sub_surf.get_rect(center=(self.WIDTH // 2, self.HEIGHT // 2 + 40))
        self.screen.blit(title_surf, tr)
        self.screen.blit(sub_surf, sr)

    def _draw(self) -> None:
        """全スプライトとUIを描画する。"""
        self.screen.fill((12, 14, 28))
        self.swarm.draw(self.screen)
        for barrier in self.barriers:
            barrier.draw(self.screen)
        self.player.draw(self.screen)
        for b in self.player_bullets:
            b.draw(self.screen)
        for b in self.enemy_bullets:
            b.draw(self.screen)
        self._draw_hud()

        if self.state == "game_over":
            self._draw_overlay("GAME OVER", "R: 最初から再開 / ESC: 終了")
        elif self.state == "game_clear":
            self._draw_overlay("GAME CLEAR!", "R: 最初から再開 / ESC: 終了")

        pygame.display.flip()

    def run(self) -> None:
        """メインループ：イベント処理・更新・描画。"""
        running = True
        while running:
            dt_ms = self.clock.tick(self.FPS)

            for event in pygame.event.get():
                if event.type == pygame.QUIT:
                    running = False
                elif event.type == pygame.KEYDOWN:
                    if event.key == pygame.K_ESCAPE:
                        running = False
                    elif event.key == pygame.K_r and self.state in ("game_over", "game_clear"):
                        self._restart_from_beginning()

            keys = pygame.key.get_pressed()
            if self.state == "playing":
                self._update_playing(keys, dt_ms)

            self._draw()

        pygame.quit()

File: web/invaders_game/js/game.js
/**
 * ファイル名：game.js
 * 説明：ゲーム全体の状態・初期化・ゲームループ・入力・衝突・レベル・勝敗判定・HUD描画を担当する。
 * Date　：2026-04-10
 * Author：Takashi Oikawa
 */

/** ゲーム全体の定数・設定 */
const GameConfig = {
  CANVAS_WIDTH: 800,
  CANVAS_HEIGHT: 600,
  /** レベル0〜3の4段階（レベル3クリアでゲームクリア） */
  MAX_LEVEL: 3,
  INITIAL_LIVES: 3,
  PLAYER_SHOOT_COOLDOWN_MS: 300,
  ENEMY_SHOOT_INTERVAL_BASE_MS: 950,
  BG_COLOR: "#0c0e1c",
};

/**
 * Canvas 上でスペースインベーダー風ゲームを動かすメインクラス。
 */
class Game {
  static STATE_PLAYING = "playing";
  static STATE_PAUSED = "paused";
  static STATE_GAME_OVER = "game_over";
  static STATE_GAME_CLEAR = "game_clear";

  /**
   * @param {HTMLCanvasElement} canvas
   */
  constructor(canvas) {
    this.canvas = canvas;
    this.ctx = /** @type {CanvasRenderingContext2D} */ (canvas.getContext("2d"));
    canvas.width = GameConfig.CANVAS_WIDTH;
    canvas.height = GameConfig.CANVAS_HEIGHT;

    this.groundY = GameConfig.CANVAS_HEIGHT - 48;
    this.invasionY = this.groundY - Player.HEIGHT - 8;

    this.score = 0;
    this.level = 0;
    this.lives = GameConfig.INITIAL_LIVES;
    this.state = Game.STATE_PLAYING;

    /** @type {Player | null} */
    this.player = null;
    /** @type {EnemySwarm | null} */
    this.swarm = null;
    /** @type {Barrier[]} */
    this.barriers = [];
    /** @type {Bullet[]} */
    this.playerBullets = [];
    /** @type {Bullet[]} */
    this.enemyBullets = [];

    this.shootCooldownMs = 0;
    this.enemyShootAccumMs = 0;

    this.leftPressed = false;
    this.rightPressed = false;
    this.spaceHeld = false;

    this._lastFrameTime = 0;
    this._rafId = 0;

    this._onKeyDown = this._onKeyDown.bind(this);
    this._onKeyUp = this._onKeyUp.bind(this);
    this._loop = this._loop.bind(this);

    window.addEventListener("keydown", this._onKeyDown);
    window.addEventListener("keyup", this._onKeyUp);

    this.spawnStage();
  }

  /** 現在レベルに応じてステージを生成する。 */
  spawnStage() {
    this.player = new Player(GameConfig.CANVAS_WIDTH, this.groundY);
    this.swarm = new EnemySwarm(GameConfig.CANVAS_WIDTH, this.level, 72);
    this.barriers = [];
    const barrierTop = this.groundY - 110;
    for (let i = 0; i < 4; i++) {
      const cx = (GameConfig.CANVAS_WIDTH * (2 * i + 1)) / 10;
      this.barriers.push(new Barrier(cx, barrierTop));
    }
    this.playerBullets = [];
    this.enemyBullets = [];
    this.shootCooldownMs = 0;
    this.enemyShootAccumMs = 0;
  }

  /** スコア・レベル・残機を初期化して最初からやり直す。 */
  restartFromBeginning() {
    this.score = 0;
    this.level = 0;
    this.lives = GameConfig.INITIAL_LIVES;
    this.state = Game.STATE_PLAYING;
    this.spawnStage();
  }

  /** @returns {number} */
  _enemyShootIntervalMs() {
    return Math.max(380, GameConfig.ENEMY_SHOOT_INTERVAL_BASE_MS - this.level * 140);
  }

  /**
   * @param {number} dtMs
   */
  _tryPlayerShoot(dtMs) {
    this.shootCooldownMs = Math.max(0, this.shootCooldownMs - dtMs);
    if (!this.spaceHeld || this.shootCooldownMs > 0 || !this.player) return;
    const pr = this.player.getBounds();
    this.playerBullets.push(
      new Bullet(pr.x + pr.width / 2, pr.y, -11, true)
    );
    this.shootCooldownMs = GameConfig.PLAYER_SHOOT_COOLDOWN_MS;
  }

  /**
   * @param {number} dtMs
   */
  _enemyTryShoot(dtMs) {
    if (!this.swarm) return;
    this.enemyShootAccumMs += dtMs;
    if (this.enemyShootAccumMs < this._enemyShootIntervalMs()) return;
    this.enemyShootAccumMs = 0;
    const shooter = this.swarm.pickRandomShooter();
    if (!shooter) return;
    const er = shooter.getBounds();
    const vy = 4 + this.level * 1.2;
    this.enemyBullets.push(
      new Bullet(er.x + er.width / 2, er.y + er.height, vy, false)
    );
  }

  _updateBullets() {
    const h = GameConfig.CANVAS_HEIGHT;
    for (const b of this.playerBullets) {
      b.update();
      b.cullIfOffScreen(h);
    }
    for (const b of this.enemyBullets) {
      b.update();
      b.cullIfOffScreen(h);
    }
    this.playerBullets = this.playerBullets.filter((b) => b.isAlive());
    this.enemyBullets = this.enemyBullets.filter((b) => b.isAlive());
  }

  _collidePlayerBullets() {
    for (const bullet of this.playerBullets) {
      if (!bullet.isAlive()) continue;
      const br = bullet.getBounds();
      for (const barrier of this.barriers) {
        if (barrier.collideBullet(br)) {
          bullet.setAlive(false);
          break;
        }
      }
      if (!bullet.isAlive()) continue;
      if (!this.swarm) continue;
      for (const enemy of this.swarm.getEnemies()) {
        if (!enemy.isAlive()) continue;
        const er = enemy.getBounds();
        if (this._rectsIntersect(br, er)) {
          enemy.setAlive(false);
          bullet.setAlive(false);
          this.score += enemy.getScoreValue();
          break;
        }
      }
    }
    this.playerBullets = this.playerBullets.filter((b) => b.isAlive());
  }

  _collideEnemyBullets() {
    for (const bullet of this.enemyBullets) {
      if (!bullet.isAlive()) continue;
      const br = bullet.getBounds();
      for (const barrier of this.barriers) {
        if (barrier.collideBullet(br)) {
          bullet.setAlive(false);
          break;
        }
      }
      if (!bullet.isAlive()) continue;
      if (this.player && this._rectsIntersect(br, this.player.getBounds())) {
        bullet.setAlive(false);
        this._onPlayerHit();
      }
    }
    this.enemyBullets = this.enemyBullets.filter((b) => b.isAlive());
  }

  /**
   * @param {{x:number,y:number,width:number,height:number}} a
   * @param {{x:number,y:number,width:number,height:number}} b
   */
  _rectsIntersect(a, b) {
    return (
      a.x < b.x + b.width &&
      a.x + a.width > b.x &&
      a.y < b.y + b.height &&
      a.y + a.height > b.y
    );
  }

  _onPlayerHit() {
    this.lives -= 1;
    if (this.player) this.player.resetPosition();
    this.enemyBullets = [];
    if (this.lives <= 0) {
      this.state = Game.STATE_GAME_OVER;
    }
  }

  _checkStageProgress() {
    if (!this.swarm || this.swarm.aliveCount() > 0) return;
    if (this.level >= GameConfig.MAX_LEVEL) {
      this.state = Game.STATE_GAME_CLEAR;
      return;
    }
    this.level += 1;
    this.spawnStage();
  }

  _checkInvasion() {
    if (!this.swarm) return;
    if (this.swarm.invasionLineReached(this.invasionY)) {
      this.state = Game.STATE_GAME_OVER;
    }
  }

  /**
   * プレイ中の1フレーム更新。
   * @param {number} dtMs
   */
  _updatePlaying(dtMs) {
    if (!this.player || !this.swarm) return;
    this.player.update(this.leftPressed, this.rightPressed);
    this._tryPlayerShoot(dtMs);
    this.swarm.update();
    this._enemyTryShoot(dtMs);
    this._updateBullets();
    this._collidePlayerBullets();
    this._collideEnemyBullets();
    if (this.state !== Game.STATE_PLAYING) return;
    this._checkInvasion();
    if (this.state !== Game.STATE_PLAYING) return;
    this._checkStageProgress();
  }

  /**
   * @param {KeyboardEvent} e
   */
  _onKeyDown(e) {
    const code = e.code;
    if (code === "ArrowLeft") {
      this.leftPressed = true;
      e.preventDefault();
    } else if (code === "ArrowRight") {
      this.rightPressed = true;
      e.preventDefault();
    } else if (code === "Space") {
      this.spaceHeld = true;
      e.preventDefault();
    } else if (code === "Escape") {
      if (
        this.state === Game.STATE_PLAYING ||
        this.state === Game.STATE_PAUSED
      ) {
        this.state =
          this.state === Game.STATE_PLAYING
            ? Game.STATE_PAUSED
            : Game.STATE_PLAYING;
      }
      e.preventDefault();
    } else if (code === "KeyR") {
      if (
        this.state === Game.STATE_GAME_OVER ||
        this.state === Game.STATE_GAME_CLEAR ||
        this.state === Game.STATE_PAUSED
      ) {
        this.restartFromBeginning();
      }
      e.preventDefault();
    }
  }

  /**
   * @param {KeyboardEvent} e
   */
  _onKeyUp(e) {
    const code = e.code;
    if (code === "ArrowLeft") this.leftPressed = false;
    else if (code === "ArrowRight") this.rightPressed = false;
    else if (code === "Space") this.spaceHeld = false;
  }

  /**
   * @param {number} timestamp
   */
  _loop(timestamp) {
    if (!this._lastFrameTime) this._lastFrameTime = timestamp;
    const dtMs = Math.min(50, timestamp - this._lastFrameTime);
    this._lastFrameTime = timestamp;

    if (this.state === Game.STATE_PLAYING) {
      this._updatePlaying(dtMs);
    }

    this._draw();
    this._rafId = requestAnimationFrame(this._loop);
  }

  /** ゲームループ開始 */
  start() {
    this._lastFrameTime = 0;
    this._rafId = requestAnimationFrame(this._loop);
  }

  /** ループ停止（ページ離脱時など） */
  stop() {
    cancelAnimationFrame(this._rafId);
    window.removeEventListener("keydown", this._onKeyDown);
    window.removeEventListener("keyup", this._onKeyUp);
  }

  _draw() {
    const ctx = this.ctx;
    const w = GameConfig.CANVAS_WIDTH;
    const h = GameConfig.CANVAS_HEIGHT;

    ctx.fillStyle = GameConfig.BG_COLOR;
    ctx.fillRect(0, 0, w, h);

    if (this.swarm) this.swarm.draw(ctx);
    for (const b of this.barriers) b.draw(ctx);
    if (this.player) this.player.draw(ctx);
    for (const b of this.playerBullets) b.draw(ctx);
    for (const b of this.enemyBullets) b.draw(ctx);

    this._drawHud();

    if (this.state === Game.STATE_PAUSED) {
      this._drawOverlay("PAUSED", "ESC: 再開 / R: 最初から");
    } else if (this.state === Game.STATE_GAME_OVER) {
      this._drawOverlay("GAME OVER", "R: 最初から再開");
    } else if (this.state === Game.STATE_GAME_CLEAR) {
      this._drawOverlay("GAME CLEAR!", "R: 最初から再開");
    }
  }

  _drawHud() {
    const ctx = this.ctx;
    ctx.font = "bold 20px sans-serif";
    ctx.fillStyle = "#e6e6e6";
    const margin = 8;
    ctx.fillText(`SCORE ${this.score}`, margin, margin + 18);
    ctx.fillText(`LIVES ${this.lives}`, margin + 220, margin + 18);
    ctx.fillText(`LEVEL ${this.level}`, margin + 400, margin + 18);
    if (this.state === Game.STATE_PLAYING) {
      ctx.font = "14px sans-serif";
      ctx.fillStyle = "#8899aa";
      ctx.fillText("ESC: 一時停止", margin + 560, margin + 18);
    }
  }

  /**
   * @param {string} title
   * @param {string} subtitle
   */
  _drawOverlay(title, subtitle) {
    const ctx = this.ctx;
    const w = GameConfig.CANVAS_WIDTH;
    const h = GameConfig.CANVAS_HEIGHT;

    ctx.fillStyle = "rgba(0,0,0,0.67)";
    ctx.fillRect(0, 0, w, h);

    ctx.font = "bold 48px sans-serif";
    ctx.fillStyle = "#ffffc8";
    const tw = ctx.measureText(title).width;
    ctx.fillText(title, (w - tw) / 2, h / 2 - 20);

    ctx.font = "18px sans-serif";
    ctx.fillStyle = "#dcdcdc";
    const sw = ctx.measureText(subtitle).width;
    ctx.fillText(subtitle, (w - sw) / 2, h / 2 + 40);
  }
}
